### PR TITLE
Add SocketTransportOption to enable/disable WaitForData

### PR DESCRIPTION
**Pull request by tmds**
This allows to opt-out of the zero-byte read that is performed to reduce memory usage for idle connections.
This read has a measurable impact on TE JSON benchmark.

cc @halter73 @davidfowl @sebastienros @adamsitnik

### CODE REVIEW

File: src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs
Comment (adamsitnik) [inline src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs]: nit: other options are exposed as properties, not fields

```suggestion
        public bool WaitForDataBeforeAllocatingBuffer { get; set; } = true;
```

**Review by adamsitnik — APPROVED** [review]
LGTM.

I confirm that making this optional can improve throughput and latency. Moreover, it's mandatory to make batching with Linux AIO work (it does not support 0 bytes reads)

File: src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs
Comment (Tratcher) [inline src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs]: Setting this to false can also increase throughput at the cost of increased memory usage.

**Review by Tratcher — COMMENTED** [review]
The ref assemblies need updating.
https://github.com/dotnet/aspnetcore/blob/master/docs/ReferenceAssemblies.md

What are the actual benchmark numbers?

**Comment by tmds** [comment]
> What are the actual benchmark numbers?

Can we run a benchmark using the Citrine infrastructure?

Should I make a change to some files to make the benchmark run with `WaitForDataBeforeAllocatingBuffer=false`?

**Comment by anurse** [comment]
FYI, this will need API review (should be quick and easy though) as it adds a new public API.

**Comment by tmds** [comment]
I set the default to `false` to ensure that is what used by benchmarks.
I'll change it back to `true` before this is merged.

**Comment by adamsitnik** [comment]
Benchmark results using Citrine:

JSON

WaitForData enabled:

```cmd
--jobs "..\Benchmarks\benchmarks.json.json" --scenario "Json"
```
```log
RequestsPerSecond:           840,803
Max CPU (%):                 100
WorkingSet (MB):             417
Avg. Latency (ms):           0.86
Startup (ms):                504
First Request (ms):          44.64
Latency (ms):                0.11
Total Requests:              12,696,202
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             3,001
Published Size (KB):         26,065
SDK:                         5.0.100-preview.2.20120.3
Runtime:                     5.0.0-preview.2.20125.16
ASP.NET Core:                5.0.0-preview.2.20126.7
```

WaitForData disabled:

```cmd
--jobs "..\Benchmarks\benchmarks.json.json" --scenario "Json"
```
```log
RequestsPerSecond:           861,744
Max CPU (%):                 100
WorkingSet (MB):             436
Avg. Latency (ms):           0.83
Startup (ms):                303
First Request (ms):          51.51
Latency (ms):                0.1
Total Requests:              13,012,363
Duration: (ms)               15,100
Socket Errors:               0
Bad Responses:               0
Build Time (ms):             4,001
Published Size (KB):         119,145
SDK:                         5.0.100-preview.2.20120.3
Runtime:                     5.0.0-preview.2.20125.16
ASP.NET Core:                5.0.0-preview.2.20126.7
```

For Plaintext there is no visible difference - both settings give me 86XYk RPS

**Comment by sebastienros** [comment]
Did you set --self-contained on both runs, or use a custom build on each run to ensure that self-contained is implicitly set? If not then the displayed versions are misleading as the sdk's shared runtimes would prevail.

**Comment by adamsitnik** [comment]
@sebastienros For the disabled case, I've compiled this dll locally and sent it using `--output-file` option. For the enabled, I've not provided any extra arguments

**Comment by sebastienros** [comment]
In the case you don't provide any argument then you need to use `--self-contained`. I force it on all the CI runs, and it's also a default on the new driver. It might not change anything as we updated the sdk recently in the aspnet repos.

File: src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs
Comment (halter73) [inline src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs]: Seems reasonable. I don't necessarily add a ton of knobs, but this one seems worthwhile. I'll approve once this gets switched back to true. I don't want to accidentally switch the default.

**Comment by tmds** [comment]
@sebastienros @halter73 where can we set `WaitForDataBeforeAllocatingBuffer` to `false` for the benchmarks?

**Comment by tmds** [comment]
> @sebastienros @halter73 where can we set WaitForDataBeforeAllocatingBuffer to false for the benchmarks?

I guess it is in https://github.com/aspnet/Benchmarks once this change has rippled through.

@anurse @halter73 is this good to merge?

**Comment by halter73** [comment]
Normally our API reviews are on Monday, but it got canceled this week. I'll try to get it through next Monday. Sorry for the delay.

**Review by halter73 — APPROVED** [review]

**Comment by sebastienros** [comment]
If the goal is only for benchmarking, you can do a local build that would use an ENV, or hard code the switch, and share the lib to use it instead.

**Comment by halter73** [comment]
This got approved via email, so no need to wait 😄

**Comment by benaadams** [comment]
As follow up to this call `GetMemory()` prior to calling `FlushAsync()` as otherwise there is contention on the Pipeline between reader and writer https://github.com/dotnet/aspnetcore/pull/22201

**Pull request merged**
